File: Trees/LC_572.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.29

class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not root:
            return False
       
        def check(p, q):
            if not p and not q:
                return True
            if not p or not q:
                return False
            return p.val == q.val and check(p.left, q.left) and check(p.right, q.right)
        
        return check(root, subRoot) or self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)

File: Trees/LC_572.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.29

class Solution {
public:
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if (root == nullptr && subRoot == nullptr){
            return true;
        }
        if (root == nullptr || subRoot == nullptr){
            return false;
        }
        return check(root, subRoot) || isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot);
    }
private:
    bool check(TreeNode* p, TreeNode* q){
        if (p == nullptr && q == nullptr){
            return true;
        }
        if (p == nullptr || q == nullptr){
            return false;
        }
        return p->val == q->val && check(p->left, q->left) && check(p->right, q->right);
    }
};